main.go

```diff
@@ -18,5 +18,194 @@
+	"strings"
+	"time"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
+	nested "github.com/antonfisher/nested-logrus-formatter"
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffetls/tlsconfig"
+	"github.com/spiffe/go-spiffe/v2/svid/x509svid"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+	"github.com/networkservicemesh/sdk/pkg/tools/log/logruslogger"
+	"github.com/networkservicemesh/sdk/pkg/tools/spiffejwt"
+	"github.com/networkservicemesh/sdk/pkg/tools/token"
+	"github.com/networkservicemesh/sdk/pkg/tools/tracing"
+)
+
+const registryAddr = "registry:5002"
+
+func getNodesHandler(w http.ResponseWriter, _ *http.Request) {
+	fmt.Println("GET /nodes requested")
+	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Access-Control-Allow-Origin", "http://localhost:3000")
+	err := json.NewEncoder(w).Encode(storageData.Nodes)
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+}
+
+func getEdgesHandler(w http.ResponseWriter, _ *http.Request) {
+	fmt.Println("GET /edges requested")
+	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Access-Control-Allow-Origin", "http://localhost:3000")
+	err := json.NewEncoder(w).Encode(storageData.Edges)
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+}
+
+func handleError(w http.ResponseWriter, err error) {
+	fmt.Println("Error: ", err)
+	http.Error(w, "Error encoding JSON", http.StatusInternalServerError)
+}
+
 func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Read polling interval env variable
+	pollingIntervalSeconds := readPollingInterval()
+
+	// Setup logger
+	log.EnableTracing(true)
+	logrus.SetFormatter(&nested.Formatter{})
+	ctx = log.WithLog(ctx, logruslogger.New(ctx, map[string]interface{}{"cmd": os.Args[:1]}))
+	logger := log.FromContext(ctx)
+
+	// Setup Spire
+	source, err2 := workloadapi.NewX509Source(ctx)
+	if err2 != nil {
+		logger.Fatalf("Error getting x509 source: %v", err2.Error())
+	}
+	var svid *x509svid.SVID
+	svid, err2 = source.GetX509SVID()
+	if err2 != nil {
+		logger.Fatalf("Error getting x509 svid: %v", err2.Error())
+	}
+	logger.Infof("sVID: %q", svid.ID)
+
+	tlsClientConfig := tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny())
+	tlsClientConfig.MinVersion = tls.VersionTLS12
+
+	configureAndRunRESTServer(logger)
+
+	// Create gRPC dial options
+	dialOptions := append(tracing.WithTracingDial(),
+		grpc.WithDefaultCallOptions(
+			grpc.WaitForReady(true),
+			grpc.PerRPCCredentials(token.NewPerRPCCredentials(spiffejwt.TokenGeneratorFunc(source, time.Duration(10)*time.Minute))),
+		),
+		grpc.WithTransportCredentials(credentials.NewTLS(tlsClientConfig)),
+	)
+
+	nsmgrs := getNsmgrsInfo(logger, dialOptions)
+
+	// Monitor Connections
+	for {
```

Comment:
The main cycle should be reworked.
Currently, we need to work fine if everything in the cluster is restarting.
To achieve it, we need to use the NSM stream API propperly. 

The strategy could be:
1. Start a goroutine that works until the main context is fine.
2. The goroutine should always try to create a connection with NSMGR.
3. For the stabilised grpc connection, we're creating two streems for nses and nsm connections and wrapping them to go channels.
4. While the main context is fine and channels are open, we simply read them.

We may also create separate goroutines for  nses,  connections,

See at the example
https://github.com/networkservicemesh/sdk/blob/main/pkg/networkservice/common/netsvcmonitor/server.go#L69-L117

Reply:
Refactoring has been done with using the event-based nsm stream api model. Could you review it again?